src/inmanta/server/services/resourceservice.py

```diff
@@ -416,6 +417,108 @@ async def mark_deployed(
             connection=connection,
         )
 
+    async def _update_deploy_state(
+        self,
+        env: data.Environment,
+        resource_id: ResourceIdStr,
+        timestamp: datetime.datetime,
+        version: int,
+        status: ResourceState,
+        message: str,
+        fail_on_error: bool,
+        is_increment_notification: bool,
+        connection: Optional[Connection] = None,
+    ) -> None:
+        """
+        Factored out the code to set a status on a resource
+
+        This excludes running `mark_done_if_done` and the agent push
+
+        Set the status of the provided resources as deployed
+        :param env: Environment to consider.
+        :param resource_id: resource to mark.
+        :param timestamp: Timestamp for the log message and the resource action entry.
+        :param version: Version of the resources to consider.
+        :param fail_on_error: When encountering an undeployable state: fail or do nothing?.
+        """
+        resource_version_id = resource_id + ",v=" + str(version)
+        logline = LogLine(
+            level=const.LogLevel.INFO,
+            msg=f"Setting {status.value} because of {message}",
+            timestamp=timestamp,
+            kwargs={"version": version},
+        )
+
+        assert status in VALID_STATES_ON_STATE_UPDATE
+        assert status not in TRANSIENT_STATES
+        async with data.Resource.get_connection(connection) as connection:
+            async with connection.transaction():
+                # validate resources
+                resource = await data.Resource.get_one(
+                    environment=env.id,
+                    resource_id=resource_id,
+                    model=version,
+                    # acquire lock on Resource before read and before lock on ResourceAction to prevent conflicts with
+                    # cascading deletes
+                    lock=data.RowLockMode.FOR_UPDATE,
```

Comment:
We do currently have deadlock issues here. This could become more of an issue now that this is under the advisory lock, so I hope our approach to address the deadlocks pays off.

Reply:
hope so too